### PR TITLE
Governor voting grab god token

### PR DESCRIPTION
**Pull request by abhishek-k-sharma-lab49**
Grab all GOD tokens when the user vote for the proposal.

### CODE REVIEW

File: contracts/governance/GODHolder.sol [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        address user
+    ) external override returns (uint256 amount) {
+        amount = _token.balanceOf(user);
+        godTokenForUsers[user] = amount;
+        _tokenService.associateTokenPublic(address(this), address(_token));
+        int256 responseCode = _tokenService.transferTokenPublic(
+            address(_token),
+            address(user),
+            address(this),
+            int64(uint64(amount))
```
Comment (shivamgarglab49) [inline contracts/governance/GODHolder.sol:46]: here we can use int256(amount) ?

File: contracts/governance/GovernorCountingSimpleInternal.sol [hunk 2]
```diff
@@ -242,35 +207,15 @@ abstract contract GovernorCountingSimpleInternal is
...
-        removeExistingVotingWeight(proposalId);
-
-        adjustIfVoterTokenBalanceChanged(proposalId, voters);
-
-        VotingWeight memory userWeight = VotingWeight(weight, support);
-        votingWeights[voter] = userWeight;
-
+        godHolder.grabTokensFromUser(voter);
+        address[] storage voters = proposalVoters[proposalId];
         voters.push(voter);
```
Comment (amandeeprehal-lab49) [inline contracts/governance/GovernorCountingSimpleInternal.sol:215]: Do you still need this?

File: integrationTest/governance/GovernorMethods.ts [hunk 3]
```diff
@@ -118,13 +122,14 @@ export default class GovernorMethods {
     const tokenId = TokenId.fromString(dex.GOD_TOKEN_ID);
     console.log(`\nInitialize contract with token  `);
     const votingDelay = 0;
-    const votingPeriod = 100; //Blocks to mint
+    const votingPeriod = 10; //Blocks to mint
```
Comment (shivamgarglab49) [inline integrationTest/governance/GovernorMethods.ts:125]: let it be 100 .

File: integrationTest/governance/governor.ts [hunk 4]
```diff
@@ -61,7 +61,11 @@ const createPair = async (
     .setGas(9000000)
     .setFunction(
       "createPair",
-      new ContractFunctionParameters().addAddress(token0).addAddress(token1)
+      new ContractFunctionParameters()
+        .addAddress(token0)
+        .addAddress(token1)
+        .addAddress(treasureId.toSolidityAddress())
```
Comment (amandeeprehal-lab49) [inline integrationTest/governance/governor.ts:67]: Why do you need this?

File: contracts/mock/ERC20Mock.sol [hunk 5]
```diff
@@ -11,6 +11,7 @@ contract ERC20Mock is IERC20 {
     uint256 userBalance;
     mapping(address => uint256) userBalances;
     bool private transferFailed;
+    bool needExact = false;
```
Comment (amandeeprehal-lab49) [inline contracts/mock/ERC20Mock.sol:14]: Pls remove this and implement what we discussed.

File: contracts/governance/GODHolder.sol [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        godTokenForUsers[user] = amount;
+        _tokenService.associateTokenPublic(address(this), address(_token));
+        int256 responseCode = _tokenService.transferTokenPublic(
+            address(_token),
+            address(user),
+            address(this),
+            int256(amount)
+        );
+        if (responseCode != HederaResponseCodes.SUCCESS) {
+            revert("GODHolder: token transfer failed to contract.");
```
Comment (amandeeprehal-lab49) [inline contracts/governance/GODHolder.sol:90]: Pls change it to require in. one line.

File: contracts/governance/GODHolder.sol [hunk 7]
```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.4;
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "../common/IERC20.sol";
+import "../common/IBaseHTS.sol";
+import "../common/hedera/HederaResponseCodes.sol";
+import "./IGODHolder.sol";
+
+contract GODHolder is IGODHolder, Initializable {
```
Comment (amandeeprehal-lab49) [inline contracts/governance/GODHolder.sol:9]: Pls write separate unit tests for this.

File: contracts/governance/GODHolder.sol [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        );
+        if (responseCode != HederaResponseCodes.SUCCESS) {
+            revert("GODHolder: token transfer failed to contract.");
+        }
+    }
+
+    function _updateActiveProposal(
+        address voter,
+        uint256 proposalId,
+        bool isAddOperation
```
[the remark this answers is not in the document]
Comment (amandeeprehal-lab49) [reply contracts/governance/GODHolder.sol:97]: https://thixalongmy.haugiang.gov.vn/media/1175/clean_code.pdf 

Go to page 41

**Review by amandeeprehal-lab49 — CHANGES_REQUESTED** [review]

File: contracts/governance/GODHolder.sol [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    function _removeAnArrayElement(
+        uint256 itemToRemove,
+        uint256[] storage items
+    ) internal {
+        uint index = items.length;
+        for (uint i = 0; i < items.length; i++) {
+            if (items[i] == itemToRemove) {
+                index = i;
```
Comment (amandeeprehal-lab49) [inline contracts/governance/GODHolder.sol:115]: You can break the loop once element is found.

File: contracts/governance/GovernorCountingSimpleInternal.sol [hunk 10]
```diff
@@ -282,7 +231,9 @@ abstract contract GovernorCountingSimpleInternal is
     ) internal virtual override {
         address creator = proposalCreators[proposalId].creator;
         returnGODToken(creator);
-        cleanup(creator);
+        address[] memory voters = proposalVoters[proposalId];
+        godHolder.removeActiveProposals(voters, proposalId);
```
Comment (amandeeprehal-lab49) [inline contracts/governance/GovernorCountingSimpleInternal.sol:235]: How about putting this call in cleanup function itself?

`godHolder.removeActiveProposals(voters, proposalId);`

Comment (amandeeprehal-lab49) [reply contracts/governance/GovernorCountingSimpleInternal.sol:235]: Its getting executed from execute and cancel.

File: test/god-holder-test.ts [hunk 11]
```diff
@@ -0,0 +1,80 @@
+import { expect } from "chai";
+import * as fs from "fs";
+
+import { loadFixture } from "@nomicfoundation/hardhat-network-helpers";
+import { ethers, upgrades } from "hardhat";
+
+describe("Governor Tests", function () {
```
Comment (amandeeprehal-lab49) [inline test/god-holder-test.ts:7]: Change names

File: test/god-holder-test.ts [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  const zeroAddress = "0x1111111000000000000000000000000000000000";
+  const precision = 100000000;
+  const total = 100 * precision;
+
+  describe("GovernorCountingSimpleInternal Upgradeable", function () {
```
Comment (amandeeprehal-lab49) [inline test/god-holder-test.ts:12]: Same

File: test/god-holder-test.ts [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  it("Verify GODHolder initialize should be failed for initialize called after instance created", async function () {
+    const { godHolder } = await loadFixture(deployFixture);
+    await expect(
+      godHolder.initialize(zeroAddress, zeroAddress)
+    ).to.revertedWith("Initializable: contract is already initialized");
+  });
+
+  it("Verify GODHolder grabtoken revert", async function () {
```
Comment (amandeeprehal-lab49) [inline test/god-holder-test.ts:59]: pls add positive cases as well.

File: test/god-holder-test.ts [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    const { godHolder, signers, tokenCont } = await loadFixture(deployFixture);
+    await expect(godHolder.revertTokensForVoter()).to.revertedWith(
+      "GODHolder: No amount for the Voter."
+    );
+    godHolder.grabTokensFromUser(signers[0].address);
+    await tokenCont.setTransaferFailed(true);
+    await expect(godHolder.revertTokensForVoter()).to.revertedWith(
+      "GODHolder: token transfer failed from contract."
+    );
+  });
```
Comment (amandeeprehal-lab49) [inline test/god-holder-test.ts:79]: Also, test for this `removeActiveProposals`

Comment (amandeeprehal-lab49) [reply test/god-holder-test.ts:79]: Idea should be to cover all external functions but when if they really simple you can skip them.

**Review by amandeeprehal-lab49 — APPROVED** [review]

**Pull request merged**